### DISCUSSION TITLE
Breaking changes in 0.9.20

### DISCUSSION THREAD

**Discussion by vlidholt** [serverpod/serverpod#401] [Announcements]
Version 0.9.20 is the most significant update to Serverpod so far. We tried to bake all breaking changes into a single update ahead of the 1.0 release.

The protocol between the client and the server is now much cleaner and conforms to more traditional JSON code - all credit to [Maximilian Fischer](https://github.com/fischerscode), who has done the larger bulk of the work on this update.

If you are running an app in production, your app will require an update to be compatible with this version of Serverpod. Therefore, you may want to continue to use an older version of Serverpod until you have a complete migration plan.

### Update the protocol
There have been some changes in the protocol specification. If referencing other modules, you will need to add the `module:MODULE_NICKNAME:` prefix.

Make sure to rerun `serverpod generate`.

### Cache
`Cache.get('key') as DataClass` becomes `Cache.get<DataClass>('key')`. Specifying the type is not optional.

### Future calls
The type of the parameter invoke requires must now be set as an generic of `FutureCall`.
New syntax:
```dart
class ExampleFutureCall extends FutureCall<Example> {
  @override
  Future<void> invoke(Session session, Example? object) async {
    // Do something interesting in the future here.
  }
}
```

### DateTime

From now on, every `DateTime` handled by Serverpod is converted to UTC. This includes times returned from the database or which have been serialized.

### Enums

Enums are now serialized and stored in the database as integers.

You can migrate your database by running:
```sql
ALTER TABLE table_name ALTER COLUMN column_name TYPE integer USING (column_name->'data'->>'index')::integer;
```

In case the enum detection does not work for your edge case you can add `enum` to your field specification.

### Runtime settings

Due to the change in how enums are serialized the `serverpod_runtime_settings` table needs to be recreated. It is safe to delete as it is automatically repopulated when the server starts. The following code will create a new table for the runtime settings:

```sql
CREATE TABLE "serverpod_runtime_settings" (
  "id" serial,
  "logSettings" json NOT NULL,
  "logSettingsOverrides" json NOT NULL,
  "logServiceCalls" boolean NOT NULL,
  "logMalformedCalls" boolean NOT NULL
);

ALTER TABLE ONLY "serverpod_runtime_settings"
  ADD CONSTRAINT serverpod_runtime_settings_pkey PRIMARY KEY (id);
```

**Comment by tusharbhambere** [comment]
💪

**Comment by realkhalife** [comment]
i got this error after upgrade to serverpod 0.9.20 :

_The method 'wrapSerializationData' isn't defined for the type 'MyObject'.
Try correcting the name to the name of an existing method, or defining a method named 'wrapSerializationData'._

**Reply by vlidholt** [reply]
Have you made sure to update the Serverpod version in all packages, including in the client?

**Reply by vlidholt** [reply]
Have you made sure to update the Serverpod version in all packages, including in the client?

**Reply by realkhalife** [reply]
i suppose my problem is related to the fact that i can't start the server.

the command :
`dart bin/main.dart`
produce : 
```
bin/main.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
import 'package:uzaapp_server/server.dart';
^
lib/server.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
import 'package:serverpod/serverpod.dart';
^
../../.pub-cache/hosted/pub.dartlang.org/serverpod-0.9.20/lib/serverpod.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
library serverpod;
^
lib/src/web/routes/root.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
import 'dart:io';
^
lib/src/generated/protocol.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
/* AUTOMATICALLY GENERATED CODE DO NOT MODIFY */
^
lib/src/generated/endpoints.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
/* AUTOMATICALLY GENERATED CODE DO NOT MODIFY */
^
../../.pub-cache/hosted/pub.dartlang.org/serverpod-0.9.20/lib/server.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
library server;
^
../../.pub-cache/hosted/pub.dartlang.org/serverpod-0.9.20/lib/relic.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
export 'src/relic/web_server.dart';
^
../../.pub-cache/hosted/pub.dartlang.org/serverpod-0.9.20/lib/database.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
library database;
^
../../.pub-cache/hosted/pub.dartlang.org/serverpod_serialization-0.9.20/lib/serverpod_serialization.dart:1:1: Error: The specified language version is too high. The highest supported language version is 2.14.
library serverpod_serialization;
^
```
Currently im on dart 2.18.3, i can't downgrade due to some packages.

So, is serverpod 0.9.20 require dart 2.14.0 ?

**Reply by yahu1031** [reply]
@realkhalife Even I use the same, it works as expected. Can you see what is the target SDK version in server.

Mine and packages version is 
```yaml
environment:
  sdk: '>=2.17.0 <3.0.0'
```

**Comment by suhaelfarouk** [comment]
Somebody Explain

SERVERPOD version: 0.9.20 mode: development time: 2022-11-24 13:16:30.435003Z
2022-11-24 13:16:31.469294Z Failed to connect to database.
FormatException: No deserialization found for type RuntimeSettings
#0      SerializationManager.deserialize (package:serverpod_serialization/src/serialization.dart:73:5)
#1      Protocol.deserialize (package:taskset_server/src/generated/protocol.dart:55:18)
#2      DatabaseConnection._formatTableRow (package:serverpod/src/database/database_connection.dart:239:45)
#3      DatabaseConnection.find (package:serverpod/src/database/database_connection.dart:178:18)
<asynchronous suspension>
#4      DatabaseConnection.findSingleRow (package:serverpod/src/database/database_connection.dart:199:18)
<asynchronous suspension>
#5      Database.findSingleRow (package:serverpod/src/database/database.dart:86:12)
<asynchronous suspension>
#6      Serverpod.start.<anonymous closure> (package:serverpod/src/server/serverpod.dart:310:13)
<asynchronous suspension>
#7      Serverpod.start (package:serverpod/src/server/serverpod.dart:299:5)
<asynchronous suspension>
#8      run (package:taskset_server/server.dart:30:3)
<asynchronous suspension>

**Reply by aldychris** [reply]
I don't know it's bug or not but I also have this issue before... 
Initially no issue when I try to run my project on 0.9.20 for first time but the 2nd time I got this issue, after that I check serverpod_runtime_settings table on my Postgres and delete the entry on that Table, after restart the server no more issue on my side.. 

Maybe the cause is because the deserialization from Protocol generated code can't found RunTimeSettingTable definition, but Im not understand why the issue is gone after I clean the table

**Reply by yahu1031** [reply]
I'm checking on it... 😇

**Reply by yahu1031** [reply]
I tried but can't reproduce it sadly.

**Reply by suhaelfarouk** [reply]
> I tried but can't reproduce it sadly.

Any workaround for now ? Should I clean table ?

**Reply by yahu1031** [reply]
yes, cleaning is only the way as of now.

**Reply by vlidholt** [reply]
I added a note about how to recreate the runtime settings table. Thanks for pointing this out!
